CefSharp.Wpf/Internals/MouseAdjustor.cs

```diff
@@ -0,0 +1,150 @@
+        /// </summary>
+        /// <param name="originalRect"></param>
+        /// <param name="viewRect"></param>
+        /// <returns>The adjusted point.</returns>
+        public System.Windows.Point UpdatePopupSizeAndPosition(Rect originalRect, Rect viewRect)
+        {
+            int x = originalRect.X,
+                prevX = originalRect.X,
+                y = originalRect.Y,
+                prevY = originalRect.Y,
+                xOffset = 0,
+                yOffset = 0;
+
+            // If popup goes outside the view, try to reposition origin
+            if (originalRect.X + originalRect.Width > viewRect.Width)
+            {
+                x = viewRect.Width - originalRect.Width;
+                xOffset = prevX - x;
+            }
+
+            if (originalRect.Y + originalRect.Height > viewRect.Height)
+            {
+                y = y - originalRect.Height - 20;
+                yOffset = prevY - y;
+            }
+
+            // If x or y became negative, move them to 0 again
+            if (x < 0)
+            {
+                x = 0;
+                xOffset = prevX;
+            }
+
+            if (y < 0)
+            {
+                y = 0;
+                yOffset = prevY;
+            }
+
+            if (x != prevX || y != prevY)
+            {
+                this.xOffset = xOffset;
+                this.yOffset = yOffset;
+
+                Rect adjustedRect = new Rect(x, y, x + originalRect.Width, y + originalRect.Height);
+
+                this.originalRect = originalRect;
+                this.adjustedRect = adjustedRect;
+
+                if (this.originalRect.Y < this.adjustedRect.Y + this.adjustedRect.Height)
+                {
+                    var newY = this.adjustedRect.Y + this.adjustedRect.Height;
+                    this.originalRect = new Rect(originalRect.X, newY, originalRect.Width, originalRect.Y + originalRect.Height - newY);
+                }
+            }
+
+            return new System.Windows.Point(x, y);
+        }
+
+        /// <summary>
+        /// Resets the offsets and original-rect.
+        /// <param name="isOpen">If the popup is open or not.</param>
+        /// </summary>
+        public void OnPopupShow(bool isOpen)
+        {
+            if (!isOpen)
+            {
+                this.xOffset = 0;
+                this.yOffset = 0;
+
+                this.originalRect = new Rect();
+                this.originalRect = new Rect();
+            }
+        }
+
+        /// <summary>
+        /// Adjusts the mouse-coordinates when the popup is visible.
+        /// </summary>
+        /// <param name="point">The original point.</param>
+        /// <returns>The adjusted point if needed, else the original point.</returns>
+        public Point GetAdjustedMouseCoords(System.Windows.Point point)
+        {
+            this.isOpen = !this.IsInsideOriginalRect(point) && IsInsideAdjustedRect(point);
```

Comment:
`isOpen` should be calculated in the other methods, not in here. By this point you should already know if the popup is open, and it's position adjusted.

Reply:
Are you sure? I looked for a solution but it was not that clean, since I had to call "IsInsideOriginalRect" and "IsInsideAdjustedRect" multiple times instead of right here. How would you envisage this in concrete terms?

Reply:
I think you've misunderstood. I would have expected this method to look something like the following (written in notepad, so not tested)

```c#
public Point GetAdjustedMouseCoords(System.Windows.Point point)
{
// Popup isn't open, or it's open and it's position was unchanged, so we don't need to perform any transform
	if (!this.isOpen)
		return new Point((int)point.X, (int)point.Y);
		
	if(!this.IsInsideOriginalRect(point) && IsInsideAdjustedRect(point))
		return new Point((int)point.X + this.xOffset, (int)point.Y + this.yOffset);
		
	return new Point((int)point.X, (int)point.Y);
}
```

- Popup is closed, so we don't need to do anything
- Popup is open, though we didn't change it's position, so no need to do anything
- Check if we need to perform a translation, not required when no over the popup

The logic can probably be cleaned up a little still. Does that make more sense now?

Reply:
Yes, thank you, I have understood that now 😄
I committed the changes.